app/cli/commandlineparser.cpp

```diff
@@ -0,0 +1,335 @@
+        if (!ok) {
+            showError(QString("Invalid %1 value: %2").arg(name, value(name)));
+        }
+        return intValue;
+    }
+
+    bool getToggleOptionValue(QString name, bool defaultValue) const
+    {
+        QRegularExpression re(QString("^(%1|no-%1)$").arg(name));
+        QStringList options = optionNames().filter(re);
+        if (options.isEmpty()) {
+            return defaultValue;
+        } else {
+            return options.last() == name;
+        }
+    }
+
+    QString getChoiceOptionValue(QString name) const
+    {
+        if(!m_Choices[name].contains(value(name))) {
+            showError(QString("Invalid %1 choice: %2").arg(name, value(name)));
+        }
+        return value(name);
+    }
+
+    QPair<int,int> getResolutionOptionValue(QString name) const
+    {
+        QRegularExpression re("^(\\d+)x(\\d+)$", QRegularExpression::CaseInsensitiveOption);
+        auto match = re.match(value(name));
+        if (!match.hasMatch()) {
+            showError(QString("Invalid %1 format: %2").arg(name, value(name)));
+        }
+        return qMakePair(match.captured(1).toInt(), match.captured(2).toInt());
+    }
+
+    void addFlagOption(QString name, QString descriptiveName)
+    {
+        addOption(QCommandLineOption(name, QString("Use %1.").arg(descriptiveName)));
+    }
+
+    void addToggleOption(QString name, QString descriptiveName)
+    {
+        addOption(QCommandLineOption(name, QString("Use %1.").arg(descriptiveName)));
+        addOption(QCommandLineOption("no-" + name, QString("Do not use %1.").arg(descriptiveName)));
+    }
+
+    void addValueOption(QString name, QString descriptiveName)
+    {
+        addOption(QCommandLineOption(name, QString("Specify %1 to use.").arg(descriptiveName), name));
+    }
+
+    void addChoiceOption(QString name, QString descriptiveName, QStringList choices)
+    {
+        addOption(QCommandLineOption(name, QString("Select %1: %2.").arg(descriptiveName, choices.join('/')), name));
+        m_Choices[name] = choices;
+    }
+
+private:
+    QMap<QString, QStringList> m_Choices;
+};
+
+GlobalCommandLineParser::GlobalCommandLineParser()
+{
+}
+
+GlobalCommandLineParser::~GlobalCommandLineParser()
+{
+}
+
+GlobalCommandLineParser::ParseResult GlobalCommandLineParser::parse(const QStringList &args)
+{
+    CommandLineParser parser;
+    parser.setupCommonOptions();
+    parser.setApplicationDescription(
+        "\n"
+        "Starts Moonlight normally if no arguments are given.\n"
+        "\n"
+        "Available actions:\n"
+        "  stream          Start streaming a game\n"
+        "\n"
+        "See 'moonlight <action> --help' for help of specific action."
+    );
+    parser.addPositionalArgument("action", "Action to execute", "<action>");
+    parser.parse(args);
+    auto posArgs = parser.positionalArguments();
+    QString action = posArgs.isEmpty() ? QString() : posArgs.first();
+
+    if (action == "") {
+        parser.handleHelpAndVersionOptions();
```

Comment:
Can you put a comment here stating that this won't return if `--help` or `--version` is specified?

Reply:
Can do